idaes/generic_models/properties/core/generic/tests/test_electrolyte_components.py

```diff
@@ -0,0 +1,420 @@
+            'HCO3-': {"type": Anion,
+                      "charge": -1,
+                      "parameter_data": {
+                           "mw": (61E-3, pyunits.kg/pyunits.mol)}},
+            'N2': {"type": Component,
+                   "parameter_data": {
+                        "mw": (28E-3, pyunits.kg/pyunits.mol)}}},
+
+        # Specifying phases
+        "phases":  {'Liq': {"type": AqueousPhase,
+                            "equation_of_state": ENRTL,
+                            "equation_of_state_options": {
+                                "pH_range": "basic"}},
+                    'Vap': {"type": VaporPhase,
+                            "equation_of_state": Ideal}},
+
+        # Set base units of measurement
+        "base_units": {"time": pyunits.s,
+                       "length": pyunits.m,
+                       "mass": pyunits.kg,
+                       "amount": pyunits.mol,
+                       "temperature": pyunits.K},
+
+        # Specifying state definition
+        "state_definition": FTPx,
+        "state_bounds": {"flow_mol": (0, 100, 1000, pyunits.mol/pyunits.s),
+                         "temperature": (273.15, 300, 500, pyunits.K),
+                         "pressure": (5e4, 1e5, 1e6, pyunits.Pa)},
+        "state_components": StateIndex.true,
+        "pressure_ref": (101325, pyunits.Pa),
+        "temperature_ref": (298.15, pyunits.K),
+
+        # Defining phase equilibria
+        }
+
+    @pytest.mark.unit
+    def test_true_component_lists_2_phase(self):
+        m = ConcreteModel()
+
+        m.fs = FlowsheetBlock(default={'dynamic': False})
+
+        m.fs.props = GenericParameterBlock(
+            default=TestNonAqueousComponents.config)
+
+        m.fs.state = m.fs.props.build_state_block(
+                [1],
+                default={"defined_state": True})
+
+        assert m.fs.props._electrolyte
+
+        assert m.fs.props.anion_set == ["HCO3-"]
+        assert m.fs.props.cation_set == ["K+"]
+        assert m.fs.props.solvent_set == ["H2O"]
+        assert m.fs.props.solute_set == ["CO2"]
+        assert m.fs.props._apparent_set == ["KHCO3"]
+        assert m.fs.props._non_aqueous_set == ["N2"]
+
+        assert m.fs.props.true_species_set == [
+            "HCO3-", "K+", "H2O", "CO2", "N2"]
+        assert m.fs.props.apparent_species_set == [
+            "H2O", "CO2", "KHCO3", "N2"]
+        assert m.fs.props.component_list == [
+            "HCO3-", "K+", "H2O", "CO2", "KHCO3", "N2"]
+
+        assert m.fs.props.true_phase_component_set == [
+            ("Liq", "HCO3-"), ("Liq", "K+"), ("Liq", "H2O"), ("Liq", "CO2"),
+            ("Vap", "H2O"), ("Vap", "CO2"), ("Vap", "N2")]
+        assert m.fs.props.apparent_phase_component_set == [
+            ("Liq", "H2O"), ("Liq", "CO2"), ("Liq", "KHCO3"),
+            ("Vap", "H2O"), ("Vap", "CO2"), ("Vap", "KHCO3"), ("Vap", "N2")]
+
+        assert m.fs.state[1].component_list is m.fs.props.true_species_set
```

Comment:
This assertion confuses me, why is the component list and true species set the same for the state block? The state block will still include the apparent species, right?

Reply:
The state block will contain both true and apparent species. However, the main framework expects there to be a `component_list` which it uses to index the component balances, etc.

If you look at the different tests, we have the ability to set the state block to use either the true or apparent species as the master component list (see line 221 for an example); this case happens to be the true component example. Thus, we want `component_list` to be the same as `true_species_list` for this case. There is another test for the reverse case where `component_list` is the same as `apparent_species_list`.

Reply:
Ok makes sense